File: 3639-zero-array-transformation-i/zero-array-transformation-i.py
class Solution:
    def isZeroArray(self, nums: List[int], queries: List[List[int]]) -> bool:
        n = len(nums)
        pre_process = [0] * (n+1)
        cur_reduction = 0
        for l,r in queries:
            pre_process[l] += 1
            pre_process[r+1] -= 1
        for i in range(n):
            cur_reduction += pre_process[i]
            if cur_reduction < nums[i]:
                return False
        return True

File: 3639-zero-array-transformation-i/zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> pre_process(n+1, 0);
        int cur_reduction = 0;
        for (vector<int>& q: queries) {
            pre_process[q[0]]++;
            pre_process[q[1]+1]--;
        }
        for (int i=0; i<n; i++) {
            cur_reduction += pre_process[i];
            if (cur_reduction < nums[i]) return false;
        }
        return true;
    }
};